### PR TITLE
Fix issues with SpinButton revealed by eslint-plugin-react-hooks

### LINKED ISSUE #5326: SpinButton: Add onChange handler
**Describe the feature that you would like added**
Make SpinButton useful by adding an OnChange handler.
As it is right now you have to figure out some way to accomplish the same effect using a combination of onIncrement, onDecrement, and onValidate which is really frustrating because it makes useful features of the control, like setting max and min values, unavailable since you have to manage the control yourself.

**What component or utility would this be added to**
SpinButton

**Additional context/screenshots**
Right now our team is trying to use a work-around of using TextField set with ```type="number"```, but this doesn't include the nice accessibility aria attribute handling of SpinButton and is missing the useful max/min value clamping.

I noticed the private _onChange method in the component has a comment saying there was a bug in React preventing onChange from being implemented. I assume that issue is no longer a blocker given that the issue is marked as closed and TextField has an onChange handler.

This affects VSTS.

Thanks!

### LINKED ISSUE #14480: SpinButton: fix issues revealed by eslint-plugin-react-hooks
#14478 enables `eslint-plugin-react-hooks` for react-next. There were a bunch of issues in SpinButton which would require a good bit of work to fix. I disabled the rule for the file for now but need to go back and actually fix things before release.

### LINKED ISSUE #15970: SpinButton doesn't work (v8-beta). increment/decrement work only once.
### Environment Information

- **Package version(s)**: 8.0.0-beta.15
- **Browser and OS versions**: Firefox 83 / Chrome 87

### Please provide a reproduction of the bug in a codepen:

https://codepen.io/k_oi/pen/abZxEwq

#### Actual behavior:

increment/decrement only once.
displayed value is 9 or 11 only.

#### Expected behavior:

can incremenet/decrement specific range (min-max).

### Others

* uncontrolled/controlled method doesn't work.
* onIncrement/onDecrement event is fired but argument value is always initlal value.

### Priorities and help requested:

Requested priority: Normal

### PR DESCRIPTION
**Pull request by czearing**
#### Pull request checklist
- [x] Fixes #14480, fixes #15970 
- [X] Include a change request file using `$ yarn change`

#### Description of changes
~~1. Updating react-internal `SpinButton` to handle onChange.~~
2. Resolving issues revealed by eslint-plugin-react-hooks. https://github.com/microsoft/fluentui/issues/14480
~~3. Updating react-internal `SpinButton.stateful.example.tsx` to show the usage of onChange.~~
4. Adding a forwardRef to `SpinButton`.

#### Focus areas to test
`SpinButton`

### CODE REVIEW

**Review by xugao — APPROVED** [review]

**Review by ecraig12345 — CHANGES_REQUESTED** [review]
Please wait on merging this until I can take a look again--I remember from the previous version of this that I had quite a bit of feedback, but I need to get some other things with v8 done ASAP and don't have time right away to review this in detail, sorry. :(

**Review by dzearing — APPROVED** [review]

**Comment by behowell** [comment]
Hi @czearing, I wanted to check if you are still planning to go forward with this PR?

It looks like it's waiting for review from @ecraig12345.

File: packages/react-internal/src/components/SpinButton/SpinButton.types.ts
Comment (ecraig12345) [inline packages/react-internal/src/components/SpinButton/SpinButton.types.ts:38]: HTMLTextAreaElement shouldn't be included in the props here. If this was copied from TextField, it's only included there because TextField actually can be an `input` or a `textarea` internally (but SpinButton can only be an `input`).

**Comment by ecraig12345** [comment]
After thinking about this and reading back through the great (and long) discussion on #5326, I think we should go ahead with the functionality fixes in this PR but revert the onChange addition for now. (But definitely get it done before the v8 release still!)

Correctly implementing onChange has a lot of nuances which weren't covered here: mainly that we should only call it on "commit" rather than on every keystroke, plus lots of edge cases. See https://hackmd.io/cLn40UjfT_KOM6r4yi2CUQ for a full proposed spec. If the proposal seems good, either I can work on it (hopefully this week) or @czearing you can give it a try if you have time.

Based on all that, I pushed some updates doing the following:
- Revert onChange addition for now
- Add a test for the case of pressing up/down arrow key or button twice
- Revert my previous example changes (see #16040 instead)
- Improve default handling
- Simplify calculation of initial value, last valid value, and precision
- Add helpers in test to reduce duplicate code (makes adding more coverage easier)
- Fix a bug where `lastValidValue` was not being updated after stepping (and add a test)

At this point the functionality issues should be fixed and I'm going to approve, but @dzearing should probably take a look as well so I'm not reviewing (partially) my own code. 😆

**Review by ecraig12345 — APPROVED** [review]

**Comment by ecraig12345** [comment]
I should also have said, thank you so much @czearing for all your work on this and other changes, and SO sorry for not reviewing sooner!!

**Comment by kazuki** [comment]
In controlledValue mode, `setValue(elementValue)` in `handleInputChange` doesn't work.
As a result, if input by the keyboard, the value will not be reflected correctly.

**Comment by ecraig12345** [comment]
@kazuki The controlled SpinButton isn't working quite properly yet. I'll be fixing that in another PR this week.

**Pull request merged**
